### PR TITLE
Refactoring of MetFrag functionality

### PR DESCRIPTION
**Pull request by jorainer**
In this PR:

- reformatting of the code to adhere to the RforMassSpec/Bioconductor coding style
- add unit tests for functions
- add initial documentation
- add needed imports/exports
- return the matching results as a `Matched` object

See also comments in the code and notes/comments below for explanations and reasoning of changes.

### CODE REVIEW

File: R/CompDbSource.R [hunk 1]
```diff
@@ -162,56 +162,110 @@ MassBankSource <- function(release = "2021.03", ...) {
...
+#'
+#' - `validAdducts`: `validAdducts(MetFragParam())` returns a `list` with the
+#'   supported adduct names of MetFrag (for positive and negative polarity).
+#'   If called on a [Spectra()] object (`validAdducts(spectra, MetFragParam())`)
+#'   it checks whether the specified adduct names (with spectra variable
+#'   `"precursorAdduct"`) are valid, given the polarity of each spectrum. The
+#'   function returns `TRUE` if all adduct definitions are valid or throws an
+#'   error.
+#'
+#' LLLLL PARAMETERS
```
Comment (jorainer) [inline R/CompDbSource.R]: I still need to complete the documentation, but that's details then.

File: R/CompDbSource.R [hunk 2]
```diff
@@ -162,56 +162,110 @@ MassBankSource <- function(release = "2021.03", ...) {
...
+
+## We Assume that the user has added the precursor adduct annotation(s) to the
+## spectra data, under the column 'precursorAdduct' e.g.:
+## load(system.file("extdata", "sps_test.Rdata",package = "MetaboAnnotation"))
+## sps_test$precursorAdduct <- "[M-H]-"
+## Define MetFrag parameters object and make connection with the online server:
+## mfParam = MetFragParam()
+## mfSrc = MetFragSource()
+##
+## testMatch <- matchSpectra(sps_test, MetFragSource(), MetFragParam())
```
Comment (jorainer) [inline R/CompDbSource.R]: just replaced `#'` here with `##` otherwise this is recognized/parsed as roxygen documentation.

File: R/CompDbSource.R [hunk 3]
```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
...
+				msg <- c(msg, paste0("Minimum number of input spectra ",
+                                     "fragments needs to be at least three"))
              msg
          })
 
-#' Default constructor function for 'MetFragParam'
+#' @rdname MetFrag
+#'
+#' @export
 MetFragParam <- function(...) {
```
Comment (jorainer) [inline R/CompDbSource.R:331]: here we need later to allow the user to specify and provide settings for MetFrag.

File: R/CompDbSource.R [hunk 4]
```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
...
-    prototype = list(url = "https://msbi.ipb-halle.de/MetFrag-deNBI/api/v1/process", 
-                     config = list(`Content-Type` = "application/json"))
+              config = "list"),
+    prototype = list(
+        url = "https://msbi.ipb-halle.de/MetFrag-deNBI/api/v1/process",
+        config = list(`Content-Type` = "application/json"))
 )
 
-#' Define the default constructor function for 'MetFragSource' and run a call to the server using a test query
-MetFragSource <- function(url = "https://msbi.ipb-halle.de/MetFrag-deNBI/api/v1/process") {
```
Comment (jorainer) [inline R/CompDbSource.R:292]: I have just reformatted here everything - also, I reduced the number of peaks here. No need to send this much data to MetFrag just to check that the server is alive.

File: R/CompDbSource.R [hunk 5]
```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
...
+        testCall <- POST(url = mfSrc@url,config = mfSrc@config,
+                         body = testQuery, encode = "json", handle)
+        ## a naive way to capture http 'success' codes 2xx:
+        if (length(testCall$status_code) &&
+            length(grep("^2", testCall$status_code))) mfSrc
+        else stop("Remote server inactive or irresponsive. ",
+                  "Returned status code: ", testCall$status_code)
+}
+
+.peaks_to_metfrag_string <- function(x) {
```
Comment (jorainer) [inline R/CompDbSource.R:442]: Simple helper function to convert a peaks matrix to a string for MetFrag. Allows us also to add a dedicated unit test for this to evaluate that it works properly for any type of input.

File: R/CompDbSource.R [hunk 6]
```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
...
+    if (nrow(x))
+        paste0(x[, "mz"], "_", x[, "intensity"], collapse = ";")
+    else character()
+}
+
+#' @importMethodsFrom ProtGenerics as.list
+.param_to_list <- function(x, skip = c("validAdducts", "THRESHFUN")) {
```
Comment (jorainer) [inline R/CompDbSource.R:449]: this converts the parameter object to a `list`. the basic function was already implemented in `ProtGenerics::as.list`.

File: R/CompDbSource.R [hunk 7]
```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
...
+#' @importFrom MetaboCoreUtils mz2mass
 setMethod(
 	"matchSpectra", signature(query = "Spectra", target = "MetFragSource",
                               param = "MetFragParam"),
-	function(query, target, param, BPPARAM = BiocParallel::SerialParam()) {
-        out <- list()
-		names(out) <- spectraData(query)$peak_id
-        mtch <- data.frame()	
+	function(query, target, param, debug = FALSE) {
+        validAdducts(query, param)
```
Comment (jorainer) [inline R/CompDbSource.R:467]: check the input `Spectra` for valid adducts before actually calling MetFrag.

File: R/CompDbSource.R [hunk 8]
```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
...
+        if (debug) handle <- verbose()
```
Comment (jorainer) [inline R/CompDbSource.R:468]: this just allows to enable/disable the verbose messages

File: R/CompDbSource.R [hunk 9]
```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
...
+        else handle <- NULL
+        out <- vector("list", length(query))
+		names(out) <- query$peak_id
+        restQuery <- .param_to_list(param)
```
Comment (jorainer) [inline R/CompDbSource.R:471]: better to do this outside of the loop. no need to call this conversion for every single spectrum.

File: R/CompDbSource.R [hunk 10]
```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
...
-					"neutralprecursormass" = as.character(neutralPrecursorMass),
-					"precursoriontype" = as.character(adduct),
-					"peakliststring" = peakListString
+				# next()
+                peakListString <- ""
+			} else
+                peakListString <- .peaks_to_metfrag_string(peaks)
+			pmz <- qi$precursorMz
+			adduct <- unlist(qi$precursorAdduct)
+			out[[i]] <- do.call(rbind, lapply(adduct, function(a) {
```
Comment (jorainer) [inline R/CompDbSource.R]: in contrast to the initial version I suggest to return a `data.frame` as a result and not a nested list. We add the `query_idx` and `adduct` to the `data.frame`, thus it's easier to identify which matches belong to each query. And we don't have to iterate/loop through the nested list later to create the result object (which will be anyway a `data.frame`).

File: R/CompDbSource.R [hunk 11]
```diff
@@ -221,196 +275,264 @@ setClass("MetFragParam",
...
-		#' ToDo:
-		#' Convert nested list to the MatchedSpectra Object?
-		#' ...
-		#' ...		
+		out <- do.call(rbind, out)
+        mtches <- data.frame(query_idx = out$query_idx,
+                             target_idx = seq_len(nrow(out)),
+                             score = out$Score,
+                             adduct = out$adduct)
+        Matched(query = query,
```
Comment (jorainer) [inline R/CompDbSource.R:519]: Result gets returned as a `Matched` object with `query` being the input `Spectra` and `target` the result from MetFrag

File: R/CompDbSource.R [hunk 12]
```diff
@@ -162,56 +162,128 @@ MassBankSource <- function(release = "2021.03", ...) {
...
+             OfflineSpectralDatabaseFile = "character",
+             ## peak matching parameters
+             fragmentpeakmatchabsolutemassdeviation = "numeric",
+             fragmentpeakmatchrelativemassdeviation = "numeric",
+             databasesearchrelativemassdeviation = "numeric",
+             ## scoring parameters
+             MetFragScoreTypes = "character",
+             MetFragScoreWeights = "numeric",
+             ## default class filtering params
 			 requirePrecursor = "logical",
```
Comment (jorainer) [inline R/CompDbSource.R:263]: do we need these parameters at all? `requirePrecursor`, `requirePrecursorPeak`, `toleranceRt`, `percentRt`?

File: R/CompDbSource.R [hunk 13]
```diff
@@ -221,196 +293,258 @@ setClass("MetFragParam",
...
-				list(
-					mfResults = mfCandidates,
-					mfServerStatus = callStatus$status
-				)
-			})
-			names(out[[i]]) <- adduct
-			Sys.sleep(1)
+					res <- read.csv(text = content(GET(callStatus$rest_url)))
+                    res$server_status <- callStatus$status
+                    ## filter based on score?
```
Comment (jorainer) [inline R/CompDbSource.R:504]: shouldn't we filter somehow the results? e.g. if the `"score"` is higher than a certain value (defined by `@THRESHFUN`?

File: R/CompDbSource.R [hunk 14]
```diff
@@ -221,196 +293,258 @@ setClass("MetFragParam",
...
-				})
-				restQuery <- c(
-					restQuery,	
-					"neutralprecursormass" = as.character(neutralPrecursorMass),
-					"precursoriontype" = as.character(adduct),
-					"peakliststring" = peakListString
+            qi <- query[i]
+			peaks <- peaksData(qi)[[1L]]
+			if (nrow(peaks) < param@minFragCount) {
+                out[[i]] <- data.frame(
```
Comment (jorainer) [inline R/CompDbSource.R:481]: Note: with this version we report ALWAYS something for a query:

- `server_status = NA`: one row in the result `data.frame` with all `NA`s.
- `server_status = "TIMEOUT"`: query was sent but a timeout was reached - or some other error from the server. one row in the result `data.frame` with all `NA`s.
- `server_status = "SUCCESS"`: query successfull, x rows in the result `data.frame`.

**Comment by jorainer** [comment]
@AharoniLab , I have updated the PR as we discussed. Now `matchSpectra` returns something for each query spectrum (even if it's only `NA` values because the server didn't reply). The user can then filter the results afterwards.

There are however some other things we should discuss:

- likely unused/unnecessary parameters in the `MetFragParam` class
- filtering of results based on the MetFrag score.

File: R/CompDbSource.R (hunk 12 above)
Comment (AharoniLab) [reply R/CompDbSource.R:263]: yes, indeed: RT information is stored on the server and will be used to increase accuracy (also with non-identical chromatographic methods) and 'requirePrecursor' set to false can perhaps be used for similarity searches of chemically related ions.

File: R/CompDbSource.R (hunk 13 above)
Comment (AharoniLab) [reply R/CompDbSource.R:504]: I need to initially let all predictions be available for testing and performance evaluations - but also later - should it not be up to the user to filter results by any arbitrary criteria?

File: R/CompDbSource.R (hunk 11 above)
Comment (AharoniLab) [reply R/CompDbSource.R:519]: Will be happy to see that work!

File: R/CompDbSource.R [hunk 15]
```diff
@@ -162,56 +162,128 @@ MassBankSource <- function(release = "2021.03", ...) {
...
+             fragmentpeakmatchabsolutemassdeviation = "numeric",
+             fragmentpeakmatchrelativemassdeviation = "numeric",
+             databasesearchrelativemassdeviation = "numeric",
+             ## scoring parameters
+             MetFragScoreTypes = "character",
+             MetFragScoreWeights = "numeric",
+             ## default class filtering params
 			 requirePrecursor = "logical",
              requirePrecursorPeak = "logical",
              THRESHFUN = "function",
```
Comment (AharoniLab) [inline R/CompDbSource.R:265]: to remove

**Pull request merged**